src/main/flight/pid.h

```diff
@@ -390,6 +390,10 @@ typedef struct pidRuntime_s {
     float feedforwardJitterFactor;
     float feedforwardBoostFactor;
 #endif
+
+#ifdef USE_ACC
+    pt3Filter_t attitudeFilter[XYZ_AXIS_COUNT];
```

Comment:
IMU algorithm is already based on / equivalent to low-pass filter. Also Euler angles are not linear. So it would be more efficient and cleaner to update filtering in IMU directly.

Reply:
I need to go through IMU once again.

On the other hand IMU/ATTITUDE is called at 500Hz freq (level modes). If I can tune filters there I still need to remove steps in the PID loop somehow. Or am I missing something?

Reply:
Sorry, I did not consider that signal is upsampled here.
The problem is similar to RC upsampling, it was discussed a lot. Low-pass filter after repeating samples (as is used here) is one option.

BTW: It is possible to reorder upsampling and leveling PID filter. Then leveling filter will be called only on 500Hz and it's result upsampled before gyro PID, saving same CPU cycles. Result should be equivalent.

Reply:
Can you please explain filter reordering in more details? I am probably missing something. Thank you.

Reply:
As long as filters are LTI (linear, time invariant), it does not mater in which order they are executed. And, with Nyquist assumptions satisfied, downsampling and upsampling does not change signal.
So it should be possible to run leveling PID at attitude update rate, then upscaling PID output before feeding it to gyro loop.

Reply:
Agree we will continue with this in another PR. Otherwise we will never merge it - we already have some new ideas.